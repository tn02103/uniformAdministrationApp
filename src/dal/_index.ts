import { ExceptionType } from "@/errors/CustomException";

export type SAErrorResponseType = {
    error: {
        exceptionType: ExceptionType,
        data?: unknown;
    }
}