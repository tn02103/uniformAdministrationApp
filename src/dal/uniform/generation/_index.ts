"use server";

import { create } from "./create";
import { markDeleted } from "./delete";
import { changeSortOrder } from "./sortOrder";
import { update } from "./update";


export const createUniformGeneration = create;
export const deleteUniformGeneration = markDeleted;
export const changeUniformGenerationSortOrder = changeSortOrder;
export const updateUniformGeneration = update;
