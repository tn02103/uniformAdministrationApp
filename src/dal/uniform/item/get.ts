import { genericSANoDataValidator, genericSAValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { Deficiency } from "@/types/deficiencyTypes";
import { uniformHistoryArgs, UniformHistroyEntry, UniformWithOwner, uniformWithOwnerArgs } from "@/types/globalUniformTypes";
import { Prisma } from "@prisma/client";
import { z } from "zod";

export type ItemLabel = {
    id: string;
    label: string;
    active: boolean;
    typeId: string;
    number: number;
    owner: {
        id: string;
        firstname: string;
        lastname: string;
    } | null;
    storageUnit: {
        id: string;
        name: string;
    } | null;
}
/**
 * Get all uniform items for the assosiation
 * @requires AuthRole.user
 * @returns List of labels for uniform items
 */
export const getItemLabels = async (): Promise<ItemLabel[]> => genericSANoDataValidator(
    AuthRole.user,
).then(([{ assosiation }]) => prisma.uniform.findMany({
    where: {
        type: {
            fk_assosiation: assosiation
        },
        recdelete: null,
    },
    include: {
        type: true,
        storageUnit: true,
        issuedEntries: {
            where: {
                dateReturned: null,
            },
            include: {
                cadet: true,
            },
        },
    },
})).then(data => data.map((item): ItemLabel => ({
    id: item.id,
    number: item.number,
    label: `${item.type.name}-${item.number}`,
    active: item.active,
    typeId: item.type.id,
    owner: item.issuedEntries.length > 0 ? {
        id: item.issuedEntries[0].cadet.id,
        firstname: item.issuedEntries[0].cadet.firstname,
        lastname: item.issuedEntries[0].cadet.lastname,
    } : null,
    storageUnit: item.storageUnit ? {
        id: item.storageUnit.id,
        name: item.storageUnit.name,
    } : null,
})));


export const getHistory = async (props: string): Promise<UniformHistroyEntry[]> => genericSAValidator(
    AuthRole.inspector,
    props,
    z.string().uuid(),
    { uniformId: props }
).then(([, uniformId]) => prisma.uniformIssued.findMany({
    where: {
        fk_uniform: uniformId,
    },
    ...uniformHistoryArgs,
}));

const getDeficienciesPropSchema = z.object({
    uniformId: z.string().uuid(),
    includeResolved: z.boolean().optional(),
});
type GetDeficienciesProps = z.infer<typeof getDeficienciesPropSchema>;

export const getDeficiencies = async (props: GetDeficienciesProps): Promise<Deficiency[]> => genericSAValidator(
    AuthRole.inspector,
    props,
    getDeficienciesPropSchema,
    { uniformId: props.uniformId }
).then(([, { uniformId, includeResolved }]) => prisma.deficiency.findMany({
    where: {
        uniformDeficiency: {
            fk_uniform: uniformId,
        },
        dateResolved: includeResolved ? undefined : null,
    },
    include: {
        type: true,
        uniformDeficiency: true,
    },
    orderBy: [
        { dateCreated: 'asc' },  // Oldest to newest
    ],
})).then((deficiencies) =>
    deficiencies.map((deficiency): Deficiency => ({
        id: deficiency.id,
        typeId: deficiency.type.id,
        typeName: deficiency.type.name,
        description: deficiency.description,
        comment: deficiency.comment,
        dateCreated: deficiency.dateCreated,
        dateUpdated: deficiency.dateUpdated,
        dateResolved: deficiency.dateResolved,
        userCreated: deficiency.userCreated,
        userUpdated: deficiency.userUpdated,
        userResolved: deficiency.userResolved,
    }))
);

const getListWithOwnerPropSchema = z.object({
    uniformTypeId: z.string().uuid(),
    orderBy: z.enum(["number", "generation", "size", "comment", "owner"]),
    asc: z.boolean(),
    filter: z.object({
        active: z.boolean().optional(),
        isReserve: z.boolean().optional(),
        issued: z.boolean().optional(),
        notIssued: z.boolean().optional(),
        inStorageUnit: z.boolean().optional(),
        generations: z.record(z.boolean()).optional(),
        sizes: z.record(z.boolean()).optional(),
    }).optional().nullable(),
});
type getListWithOwnerProps = z.infer<typeof getListWithOwnerPropSchema>;

/**
 * used to get the data for the uniformList-Overview 
 * returns all data of unifrom, the owner is returned in the object of issuedEntries as a cadetDescription. 
 * @requires AuthRole.user
 * @param uniformTypeId 
 * @param orderBy number| generation | size | comment | owner
 * @param asc 
 * @param filter filterObject of uniformFilter Pannel
 * @returns UniformWithOwner[]
 */
export const getListWithOwner = async (props: getListWithOwnerProps): Promise<UniformWithOwner[]> => genericSAValidator(
    AuthRole.user,
    props,
    getListWithOwnerPropSchema,
    { uniformTypeId: props.uniformTypeId }
).then(async ([{ }, { uniformTypeId, orderBy, asc, filter }]) => {
    const sqlFilter: Prisma.UniformFindManyArgs["where"] = {};
    const generationFilter = filter?.generations ? Object.entries(filter.generations).filter(([, value]) => value).map(([key,]) => key) : [];
    const sizeFilter = filter?.sizes ? Object.entries(filter.sizes).filter(([, value]) => value).map(([key,]) => key) : [];
    const andConditions: Prisma.UniformWhereInput[] = [];
    if (!filter) {
        sqlFilter["active"] = true;
    } else {
        if (!filter.isReserve) {
            sqlFilter["active"] = true;
        } else if (!filter.active) {
            sqlFilter["active"] = false;
        }

        if (!filter.issued || !filter.notIssued || !filter.inStorageUnit) {
            const orConditions: Prisma.UniformWhereInput[] = [];
            if (filter.issued) {
                orConditions.push({
                    issuedEntries: {
                        some: {
                            dateReturned: null,
                        }
                    }
                });
            }
            if (filter.notIssued) {
                orConditions.push({
                    issuedEntries: {
                        none: {
                            dateReturned: null,
                        }
                    }
                });
            }
            if (filter.inStorageUnit) {
                orConditions.push({
                    storageUnit: {
                        isNot: null,
                    }
                });
            }
            andConditions.push({
                OR: orConditions,
            });
        }
    }
    if (generationFilter.length > 0) {
        andConditions.push({
            OR: [
                { generation: { id: { in: generationFilter } } },
                generationFilter.includes("null") ? { generation: null } : {}
            ]
        });
    }
    if (sizeFilter.length > 0) {
        andConditions.push({
            OR: [
                { size: { id: { in: sizeFilter } } },
                sizeFilter.includes("null") ? { size: null } : {}
            ]
        });
    }

    let sortOrder: Prisma.UniformOrderByWithRelationInput[];
    const ascString = asc ? "asc" : "desc";
    switch (orderBy) {
        case "generation":
            sortOrder = [
                { generation: { sortOrder: ascString } },
                { size: { sortOrder: ascString } },
                { number: ascString },
            ];
            break;
        case "size":
            sortOrder = [
                { size: { sortOrder: ascString } },
                { generation: { sortOrder: ascString } },
                { number: ascString },
            ];
            break;
        case "comment":
            sortOrder = [
                { comment: ascString },
                { number: ascString },
            ];
            break;
        case "owner":
            sortOrder = [
                { number: ascString }
            ];
            break;
        default:
            sortOrder = [{
                number: ascString,
            }];
            break;
    }

    const data = await prisma.uniform.findMany({
        ...uniformWithOwnerArgs,
        where: {
            ...sqlFilter,
            fk_uniformType: uniformTypeId,
            recdelete: null,
            AND: andConditions.length > 0 ? andConditions : undefined,
        },
        orderBy: sortOrder
    });

    if (orderBy !== "owner") {
        return data;
    }

    return data.sort((a, b) => {
        const returnValue = (value: number) => asc ? value : -value;

        // Sort by owner first
        if ((a.issuedEntries.length > 0) && (b.issuedEntries.length > 0)) {
            const nameA = a.issuedEntries[0].cadet.lastname + a.issuedEntries[0].cadet.firstname;
            const nameB = b.issuedEntries[0].cadet.lastname + b.issuedEntries[0].cadet.firstname;
            if (nameA === nameB) {
                return returnValue(a.number - b.number);
            } else {
                return returnValue(nameA.localeCompare(nameB));
            }
        }
        if (a.issuedEntries.length > 0) {
            return returnValue(-1);
        }
        if (b.issuedEntries.length > 0) {
            return returnValue(1);
        }
        // If no owner, sort by storage unit
        if (a.storageUnit && b.storageUnit) {
            if (a.storageUnit.name === b.storageUnit.name) {
                return returnValue(a.number - b.number);
            } else {
                return returnValue(a.storageUnit.name.localeCompare(b.storageUnit.name));
            }
        }
        if (a.storageUnit) {
            return returnValue(-1);
        }
        if (b.storageUnit) {
            return returnValue(1);
        }

        // If no owner and no storage unit, sort by number
        return returnValue(a.number - b.number);
    });
});


