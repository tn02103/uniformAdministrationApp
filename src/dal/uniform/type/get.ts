"use server"
import { genericSAValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { uniformTypeArgs } from "@/types/globalUniformTypes";
import { z } from "zod";

export const getUniformType = (props: string) => genericSAValidator(
    AuthRole.user,
    props,
    z.string().uuid(),
).then(([typeId, { assosiation }]) => prisma.uniformType.findUnique({
    where: {
        id: typeId,
        fk_assosiation: assosiation,
    },
    ...uniformTypeArgs
}));


