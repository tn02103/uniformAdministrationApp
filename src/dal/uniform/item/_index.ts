"use server";

import { create } from "./create";
import { markDeleted } from "./delete";
import { getItemLabels } from "./get";
import { issue, IssuePropType } from "./issue";
import { returnItem } from "./return";
import { update } from "./update";


export const issueUniformItem = issue;
export const createUniformItems = create;
export const deleteUniformItem = markDeleted;
export const updateUniformItem = update;
export const returnUniformItem = returnItem;
export const getUniformItemLabels = getItemLabels;

export type IssueUniformItemDataType = IssuePropType;
