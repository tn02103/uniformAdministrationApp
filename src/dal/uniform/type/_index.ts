"use server";

import { create } from "./create";
import { markDeleted } from "./delete";
import { getType, getList } from "./get";
import { changeSortOrder } from "./sortOrder";
import { update } from "./update";


export const createUniformType = create;
export const deleteUniformType = markDeleted;
export const getUniformType = getType;
export const getUniformTypeList = getList;
export const changeUniformTypeSortOrder = changeSortOrder;
export const updateUniformType = update;
