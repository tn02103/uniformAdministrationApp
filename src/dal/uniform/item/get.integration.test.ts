import { cleanData, cleanDataV2 } from "@/dal/_helper/testHelper";
import { StaticData } from "../../../../tests/_playwrightConfig/testData/staticDataLoader";
import { getUniformItemCountByType, getUniformItemLabels, getUniformListWithOwner } from "./_index";
import { getDeficiencies, getHistory } from "./get";

const { ids, data } = new StaticData(0);

it('should return a list of uniform history entries', async () => {
    const uniformId = ids.uniformIds[0][86];
    const result = await getHistory(uniformId);
    expect(result).toBeDefined();
    expect(result.length).toBeGreaterThan(0);
    cleanData(result, ["id", "cadet.id", "cadet.recdelete"]);
    expect(result).toMatchSnapshot();
    expect(result[0].cadet.recdelete).toBeUndefined();
    expect(result[1].cadet.recdelete).toBeUndefined();
    expect(result[2].cadet.recdelete).toBeUndefined();
    expect(result[3].cadet.recdelete).not.toBeNull();
});
it('should return a list of defficiencies', async () => {
    const result = await getDeficiencies({ uniformId: ids.uniformIds[0][46] });

    expect(result).toBeDefined();
    expect(result.length).toBe(2);
    expect(result[0].dateCreated?.getTime()).toBeLessThan(result[1].dateCreated!.getTime());
    expect(cleanDataV2(result)).toMatchSnapshot()
});
it('should return a list of defficiencies with resolved', async () => {
    const result = await getDeficiencies({ uniformId: ids.uniformIds[0][46], includeResolved: true });

    expect(result).toBeDefined();
    expect(result.length).toBe(4);
    expect(result[0].dateCreated?.getTime()).toBeLessThan(result[1].dateCreated!.getTime());
    expect(result[1].dateCreated?.getTime()).toBeLessThan(result[2].dateCreated!.getTime());
    expect(result[2].dateCreated?.getTime()).toBeLessThan(result[3].dateCreated!.getTime());
    expect(cleanDataV2(result)).toMatchSnapshot()
});

describe('getCountByType', () => {
    it('should return the count of uniforms for a specific type', async () => {
        const uniformTypeId = ids.uniformTypeIds[0];
        const count = data.uniformList.filter(u => (u.fk_uniformType === uniformTypeId) && !u.recdelete).length;

        const result = await getUniformItemCountByType(uniformTypeId);

        expect(result).toBeDefined();
        expect(typeof result).toBe('number');
        expect(result).toBe(count);
    });
});

describe('getItemLabels', () => {
    it('should return a list of uniform item labels', async () => {
        const result = await getUniformItemLabels();

        expect(result).toBeDefined();
        expect(Array.isArray(result)).toBe(true);
        expect(result.length).toBeGreaterThan(0);

        // Check structure of first item
        const firstItem = result[0];
        expect(firstItem).toHaveProperty('id');
        expect(firstItem).toHaveProperty('label');
        expect(firstItem).toHaveProperty('number');
        expect(firstItem).toHaveProperty('active');
        expect(firstItem).toHaveProperty('type');
        expect(firstItem.type).toHaveProperty('id');
        expect(firstItem.type).toHaveProperty('name');
        expect(firstItem.type).toHaveProperty('acronym');

        // Clean sensitive data for snapshot
        cleanData(result, ["id", "type.id", "owner?.id", "storageUnit?.id"]);
        expect(result.slice(0, 10)).toMatchSnapshot(); // Take first 10 for manageable snapshots
    });

    it('should include owner information when uniform is issued', async () => {
        const result = await getUniformItemLabels();
        const issuedItems = result.filter(item => item.owner !== null);

        expect(issuedItems.length).toBeGreaterThan(0);
        const issuedItem = issuedItems[0];
        expect(issuedItem.owner).toHaveProperty('id');
        expect(issuedItem.owner).toHaveProperty('firstname');
        expect(issuedItem.owner).toHaveProperty('lastname');
    });

    it('should include storage unit information when uniform is in storage', async () => {
        const result = await getUniformItemLabels();
        const storageItems = result.filter(item => item.storageUnit !== null);

        // At least check that the filter works correctly
        expect(storageItems.every(item => item.storageUnit !== null)).toBe(true);

        // If there are storage items, check their structure
        storageItems.forEach(storageItem => {
            expect(storageItem.storageUnit).toHaveProperty('id');
            expect(storageItem.storageUnit).toHaveProperty('name');
        });
    });
});

describe('<UniformItem> getListWithOwner', () => {
    it('sorts by number ascending', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'number',
            asc: true,
        });
        expect(result.length).toBeGreaterThan(1);
        for (let i = 1; i < result.length; i++) {
            expect(result[i].number).toBeGreaterThanOrEqual(result[i - 1].number);
        }
    });

    it('sorts by number descending', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'number',
            asc: false,
        });
        expect(result.length).toBeGreaterThan(1);
        for (let i = 1; i < result.length; i++) {
            expect(result[i].number).toBeLessThanOrEqual(result[i - 1].number);
        }
    });

    it('sorts by generation ascending', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'generation',
            asc: true,
        });
        expect(result.length).toBeGreaterThan(1);
        for (let i = 1; i < result.length; i++) {
            const prev = result[i - 1].generation?.name ?? 'ZZ';
            const curr = result[i].generation?.name ?? 'ZZ';
            expect(curr.localeCompare(prev)).toBeGreaterThanOrEqual(0);
        }
    });

    it('sorts by generation descending', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'generation',
            asc: false,
        });
        expect(result.length).toBeGreaterThan(1);
        for (let i = 1; i < result.length; i++) {
            const prev = result[i - 1].generation?.name ?? 'ZZ';
            const curr = result[i].generation?.name ?? 'ZZ';
            expect(curr.localeCompare(prev)).toBeLessThanOrEqual(0);
        }
    });

    it('sorts by size ascending', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'size',
            asc: true,
        });
        expect(result.length).toBeGreaterThan(1);
        for (let i = 1; i < result.length; i++) {
            const prev = result[i - 1].size?.name ?? 'ZZ';
            const curr = result[i].size?.name ?? 'ZZ';
            expect(curr.localeCompare(prev, undefined, { numeric: true })).toBeGreaterThanOrEqual(0);
        }
    });

    it('sorts by size descending', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'size',
            asc: false,
        });
        expect(result.length).toBeGreaterThan(1);
        for (let i = 1; i < result.length; i++) {
            const prev = result[i - 1].size?.name ?? 'ZZ';
            const curr = result[i].size?.name ?? 'ZZ';
            expect(curr.localeCompare(prev, undefined, { numeric: true })).toBeLessThanOrEqual(0);
        }
    });

    it('sorts by comment ascending', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'comment',
            asc: true,
        });
        expect(result.length).toBeGreaterThan(1);
        for (let i = 1; i < result.length; i++) {
            const prev = result[i - 1].comment ?? 'ZZ';
            const curr = result[i].comment ?? 'ZZ';
            expect(curr.localeCompare(prev)).toBeGreaterThanOrEqual(0);
        }
    });

    it('sorts by comment descending', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'comment',
            asc: false,
        });
        expect(result.length).toBeGreaterThan(1);
        for (let i = 1; i < result.length; i++) {
            const prev = result[i - 1].comment ?? 'ZZ';
            const curr = result[i].comment ?? 'ZZ';
            expect(curr.localeCompare(prev)).toBeLessThanOrEqual(0);
        }
    });

    it('sorts by owner ascending', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'owner',
            asc: true,
        });
        expect(result.length).toBeGreaterThan(1);

        // Partition the result
        const withOwner = result.filter(u => u.issuedEntries.length > 0);
        const withStorageUnit = result.filter(u => u.issuedEntries.length === 0 && u.storageUnit);
        const rest = result.filter(u => u.issuedEntries.length === 0 && !u.storageUnit);

        // 1. All with owner come first, sorted by lastname, firstname
        for (let i = 1; i < withOwner.length; i++) {
            const prev = withOwner[i - 1].issuedEntries[0].cadet!;
            const curr = withOwner[i].issuedEntries[0].cadet!;
            const cmpLast = `${curr.lastname}${curr.firstname}`.localeCompare(`${prev.lastname}${prev.firstname}`);
            expect(cmpLast).toBeGreaterThanOrEqual(0);
        }
        // 2. All with storageUnit come next, sorted by storageUnit.name
        for (let i = 1; i < withStorageUnit.length; i++) {
            const prev = withStorageUnit[i - 1].storageUnit!;
            const curr = withStorageUnit[i].storageUnit!;
            expect(curr.name.localeCompare(prev.name)).toBeGreaterThanOrEqual(0);
        }
        // 3. The rest are sorted by number
        for (let i = 1; i < rest.length; i++) {
            expect(rest[i].number).toBeGreaterThanOrEqual(rest[i - 1].number);
        }
        // 4. The order of the groups
        expect(result.slice(0, withOwner.length)).toEqual(withOwner);
        expect(result.slice(withOwner.length, withOwner.length + withStorageUnit.length)).toEqual(withStorageUnit);
        expect(result.slice(withOwner.length + withStorageUnit.length)).toEqual(rest);
    });

    it('sorts by owner descending', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'owner',
            asc: false,
        });
        expect(result.length).toBeGreaterThan(1);

        // Partition the result
        const withOwner = result.filter(u => u.issuedEntries.length > 0);
        const withStorageUnit = result.filter(u => u.issuedEntries.length === 0 && u.storageUnit);
        const rest = result.filter(u => u.issuedEntries.length === 0 && !u.storageUnit);

        // 1. All with owner come first, sorted by lastname, firstname (desc)
        for (let i = 1; i < withOwner.length; i++) {
            const prev = withOwner[i - 1].issuedEntries[0].cadet!;
            const curr = withOwner[i].issuedEntries[0].cadet!;
            const cmpLast = `${curr.lastname}${curr.firstname}`.localeCompare(`${prev.lastname}${prev.firstname}`);
            expect(cmpLast).toBeLessThanOrEqual(0);
        }
        // 2. All with storageUnit come next, sorted by storageUnit.name (desc)
        for (let i = 1; i < withStorageUnit.length; i++) {
            const prev = withStorageUnit[i - 1].storageUnit!;
            const curr = withStorageUnit[i].storageUnit!;
            expect(curr.name.localeCompare(prev.name)).toBeLessThanOrEqual(0);
        }
        // 3. The rest are sorted by number (desc)
        for (let i = 1; i < rest.length; i++) {
            expect(rest[i].number).toBeLessThanOrEqual(rest[i - 1].number);
        }
        // 4. The order of the groups
        expect(result.slice(0, rest.length)).toEqual(rest);
        expect(result.slice(rest.length, rest.length + withStorageUnit.length)).toEqual(withStorageUnit);
        expect(result.slice(rest.length + withStorageUnit.length)).toEqual(withOwner);
    });

    it('filters by generations', async () => {
        const generationId = ids.uniformGenerationIds[0];
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'number',
            asc: true,
            filter: {
                generations: {
                    [ids.uniformGenerationIds[0]]: true,
                    [ids.uniformGenerationIds[1]]: false,
                }
            }
        });
        expect(result.length).toBeGreaterThan(1);
        expect(result.every(u => u.generation?.id === generationId)).toBe(true);
    });

    it('filters by sizes', async () => {
        const sizeId = ids.sizeIds[1];
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'number',
            asc: true,
            filter: {
                sizes: {
                    [ids.sizeIds[0]]: false,
                    [ids.sizeIds[1]]: true,
                },
            },
        });
        expect(result.length).toBeGreaterThan(1);
        expect(result.every(u => u.size?.id === sizeId)).toBe(true);
    });

    it('filters active uniforms', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'number',
            asc: true,
            filter: {
                active: true,
                isReserve: false,
            }
        });
        expect(result.length).toBeGreaterThan(1);
        expect(result.every(u => u.active === true)).toBe(true);
    });

    it('filters reserve uniforms', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'number',
            asc: true,
            filter: {
                isReserve: true,
                active: false,
            }
        });
        expect(result.length).toBeGreaterThan(1);
        expect(result.every(u => u.active === false)).toBe(true);
    });

    it('filters issued uniforms', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'number',
            asc: true,
            filter: {
                issued: true,
                notIssued: false,
                inStorageUnit: false,
            }
        });
        expect(result.length).toBeGreaterThan(1);
        expect(result.every(u => u.issuedEntries.length > 0)).toBe(true);
    });

    it('filters not issued uniforms', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'number',
            asc: true,
            filter: {
                issued: false,
                notIssued: true,
                inStorageUnit: false,
            }
        });
        expect(result.length).toBeGreaterThan(1);
        expect(result.every(u => u.issuedEntries.length === 0)).toBe(true);
    });

    it('filters in storage unit', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'number',
            asc: true,
            filter: {
                active: true,
                isReserve: true,
                issued: false,
                notIssued: false,
                inStorageUnit: true,
            }
        });
        expect(result.length).toBeGreaterThan(1);
        expect(result.every(u => u.storageUnit?.id)).toBe(true);
    });

    it('filters in storage unit or issued', async () => {
        const result = await getUniformListWithOwner({
            uniformTypeId: ids.uniformTypeIds[0],
            orderBy: 'number',
            asc: true,
            filter: {
                active: true,
                isReserve: true,
                issued: true,
                notIssued: false,
                inStorageUnit: true,
            }
        });
        expect(result.length).toBeGreaterThan(1);
        expect(result.every(u => u.storageUnit?.id || u.issuedEntries.length > 0)).toBe(true);
    });
});
