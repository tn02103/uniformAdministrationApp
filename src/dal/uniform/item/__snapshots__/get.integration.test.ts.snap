// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getItemLabels should return a list of uniform item labels 1`] = `
[
  {
    "active": true,
    "label": "Typ1-1100",
    "number": 1100,
    "owner": {
      "firstname": "Maik",
      "id": "0081c5d0-713b-4c4f-a2c3-81ec497acb29",
      "lastname": "Finkel",
    },
    "storageUnit": null,
    "type": {
      "acronym": "AA",
      "name": "Typ1",
    },
  },
  {
    "active": true,
    "label": "Typ1-1101",
    "number": 1101,
    "owner": {
      "firstname": "Maik",
      "id": "0081c5d0-713b-4c4f-a2c3-81ec497acb29",
      "lastname": "Finkel",
    },
    "storageUnit": null,
    "type": {
      "acronym": "AA",
      "name": "Typ1",
    },
  },
  {
    "active": true,
    "label": "Typ1-1102",
    "number": 1102,
    "owner": {
      "firstname": "Maik",
      "id": "0081c5d0-713b-4c4f-a2c3-81ec497acb29",
      "lastname": "Finkel",
    },
    "storageUnit": null,
    "type": {
      "acronym": "AA",
      "name": "Typ1",
    },
  },
  {
    "active": true,
    "label": "Typ1-1103",
    "number": 1103,
    "owner": null,
    "storageUnit": null,
    "type": {
      "acronym": "AA",
      "name": "Typ1",
    },
  },
  {
    "active": true,
    "label": "Typ1-1104",
    "number": 1104,
    "owner": null,
    "storageUnit": null,
    "type": {
      "acronym": "AA",
      "name": "Typ1",
    },
  },
  {
    "active": false,
    "label": "Typ1-1105",
    "number": 1105,
    "owner": null,
    "storageUnit": null,
    "type": {
      "acronym": "AA",
      "name": "Typ1",
    },
  },
  {
    "active": false,
    "label": "Typ1-1106",
    "number": 1106,
    "owner": null,
    "storageUnit": null,
    "type": {
      "acronym": "AA",
      "name": "Typ1",
    },
  },
  {
    "active": false,
    "label": "Typ1-1107",
    "number": 1107,
    "owner": null,
    "storageUnit": null,
    "type": {
      "acronym": "AA",
      "name": "Typ1",
    },
  },
  {
    "active": false,
    "label": "Typ1-1108",
    "number": 1108,
    "owner": null,
    "storageUnit": {
      "id": "003984d9-17ac-423d-b5d9-8f9473c225d5",
      "name": "Kiste 01",
    },
    "type": {
      "acronym": "AA",
      "name": "Typ1",
    },
  },
  {
    "active": false,
    "label": "Typ1-1109",
    "number": 1109,
    "owner": null,
    "storageUnit": {
      "id": "003984d9-17ac-423d-b5d9-8f9473c225d5",
      "name": "Kiste 01",
    },
    "type": {
      "acronym": "AA",
      "name": "Typ1",
    },
  },
]
`;

exports[`should return a list of defficiencies 1`] = `
[
  {
    "comment": "Uniform Deficiency Sven Keller Unresolved",
    "dateCreated": {},
    "dateResolved": null,
    "dateUpdated": {},
    "description": "Typ1-1146",
    "id": "c41263-aa86-489b-a4cf-536062ac413b",
    "typeId": "8ddf5b-3238-4105-9749-70b86a6c1e00",
    "typeName": "Uniform",
    "userCreated": "test2",
    "userResolved": null,
    "userUpdated": "test3",
  },
  {
    "comment": "Broken Uniform Deficiency Unresolved",
    "dateCreated": {},
    "dateResolved": null,
    "dateUpdated": {},
    "description": "Typ1-1146",
    "id": "d3a7f8-0a9e-4b6c-bd2e-0f1a5b8c3d7e",
    "typeId": "ab6b52-309a-4355-bd47-36234dc370ec",
    "typeName": "Uniform Broken",
    "userCreated": "test4",
    "userResolved": null,
    "userUpdated": "test4",
  },
]
`;

exports[`should return a list of defficiencies with resolved 1`] = `
[
  {
    "comment": "Uniform Deficiency Sven Keller Unresolved",
    "dateCreated": {},
    "dateResolved": null,
    "dateUpdated": {},
    "description": "Typ1-1146",
    "id": "c41263-aa86-489b-a4cf-536062ac413b",
    "typeId": "8ddf5b-3238-4105-9749-70b86a6c1e00",
    "typeName": "Uniform",
    "userCreated": "test2",
    "userResolved": null,
    "userUpdated": "test3",
  },
  {
    "comment": "Uniform Deficiency Sven Keller Resolved",
    "dateCreated": {},
    "dateResolved": {},
    "dateUpdated": {},
    "description": "Typ1-1146",
    "id": "69d747-c9b1-4a63-9795-d2c51939fc18",
    "typeId": "8ddf5b-3238-4105-9749-70b86a6c1e00",
    "typeName": "Uniform",
    "userCreated": "test4",
    "userResolved": "test4",
    "userUpdated": "test4",
  },
  {
    "comment": "Broken Uniform Deficiency Resolved",
    "dateCreated": {},
    "dateResolved": {},
    "dateUpdated": {},
    "description": "Typ1-1146",
    "id": "c0f1a2-4b5d-4e8f-9a3b-6c7d0e1f2b5c",
    "typeId": "ab6b52-309a-4355-bd47-36234dc370ec",
    "typeName": "Uniform Broken",
    "userCreated": "test4",
    "userResolved": "test4",
    "userUpdated": "test4",
  },
  {
    "comment": "Broken Uniform Deficiency Unresolved",
    "dateCreated": {},
    "dateResolved": null,
    "dateUpdated": {},
    "description": "Typ1-1146",
    "id": "d3a7f8-0a9e-4b6c-bd2e-0f1a5b8c3d7e",
    "typeId": "ab6b52-309a-4355-bd47-36234dc370ec",
    "typeName": "Uniform Broken",
    "userCreated": "test4",
    "userResolved": null,
    "userUpdated": "test4",
  },
]
`;

exports[`should return a list of uniform history entries 1`] = `
[
  {
    "cadet": {
      "firstname": "Marie",
      "lastname": "Becker",
    },
    "dateIssued": 2023-08-16T00:00:00.000Z,
    "dateReturned": null,
  },
  {
    "cadet": {
      "firstname": "Antje",
      "lastname": "Fried",
    },
    "dateIssued": 2023-08-11T00:00:00.000Z,
    "dateReturned": 2023-08-15T00:00:00.000Z,
  },
  {
    "cadet": {
      "firstname": "Christina",
      "lastname": "Faber",
    },
    "dateIssued": 2023-08-06T00:00:00.000Z,
    "dateReturned": 2023-08-10T00:00:00.000Z,
  },
  {
    "cadet": {
      "firstname": "Simone",
      "lastname": "Osterhagen",
    },
    "dateIssued": 2023-08-01T00:00:00.000Z,
    "dateReturned": 2023-08-05T00:00:00.000Z,
  },
]
`;
